File: app/src/main/java/com/example/engine/SoundEngine.kt
package com.example.engine

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.Random
import kotlin.math.*

class SoundEngine {
    private val scope = CoroutineScope(Dispatchers.Default)
    private val sampleRate = 22050
    private val random = Random()
    var isMuted = false

    private fun playPcm(generator: (Int, Int) -> ShortArray) {
        if (isMuted) return
        scope.launch {
            try {
                val samples = generator(sampleRate, random.nextInt())
                if (samples.isEmpty()) return@launch

                val audioTrack = AudioTrack.Builder()
                    .setAudioAttributes(
                        AudioAttributes.Builder()
                            .setUsage(AudioAttributes.USAGE_GAME)
                            .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                            .build()
                    )
                    .setAudioFormat(
                        AudioFormat.Builder()
                            .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                            .setSampleRate(sampleRate)
                            .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                            .build()
                    )
                    .setBufferSizeInBytes(samples.size * 2)
                    .setTransferMode(AudioTrack.MODE_STATIC)
                    .build()

                audioTrack.write(samples, 0, samples.size)
                audioTrack.play()
                // Wait and release
                kotlinx.coroutines.delay((samples.size * 1000L / sampleRate) + 50)
                audioTrack.release()
            } catch (e: Exception) {
                // Ignore audio errors gracefully
            }
        }
    }

    fun playPlace() {
        playPcm { rate, _ ->
            val duration = (rate * 0.08).toInt()
            val data = ShortArray(duration)
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val freq = 240.0 - (t / 0.08) * 120.0 // pitch drop pop
                val env = (1.0 - t / 0.08).pow(1.5)
                val sample = sin(2.0 * Math.PI * freq * t) * env
                data[i] = (sample * 16000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }

    fun playDig() {
        playPcm { rate, _ ->
            val duration = (rate * 0.06).toInt()
            val data = ShortArray(duration)
            var last = 0.0
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val env = (1.0 - t / 0.06)
                val noise = random.nextDouble() * 2.0 - 1.0
                last = (last * 0.7 + noise * 0.3) // low pass filter
                data[i] = (last * env * 14000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }

    fun playBreak() {
        playPcm { rate, _ ->
            val duration = (rate * 0.15).toInt()
            val data = ShortArray(duration)
            var last = 0.0
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val env = (1.0 - t / 0.15).pow(0.8)
                val noise = random.nextDouble() * 2.0 - 1.0
                last = (last * 0.6 + noise * 0.4)
                val sub = sin(2.0 * Math.PI * (120.0 - t * 400.0) * t) * 0.5
                val sample = (last * 0.8 + sub * 0.2) * env
                data[i] = (sample * 22000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }

    fun playStep() {
        playPcm { rate, _ ->
            val duration = (rate * 0.05).toInt()
            val data = ShortArray(duration)
            var last = 0.0
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val env = (1.0 - t / 0.05)
                val noise = random.nextDouble() * 2.0 - 1.0
                last = (last * 0.8 + noise * 0.2)
                data[i] = (last * env * 8000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }

    fun playJump() {
        playPcm { rate, _ ->
            val duration = (rate * 0.12).toInt()
            val data = ShortArray(duration)
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val freq = 160.0 + (t / 0.12) * 200.0 // pitch swoop up
                val env = sin((t / 0.12) * Math.PI)
                val sample = sin(2.0 * Math.PI * freq * t) * env
                data[i] = (sample * 12000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }

    fun playHurt() {
        playPcm { rate, _ ->
            val duration = (rate * 0.2).toInt()
            val data = ShortArray(duration)
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val freq = 280.0 - (t / 0.2) * 160.0 // pitch drop "oof"
                val env = (1.0 - t / 0.2).pow(1.2)
                val square = if (sin(2.0 * Math.PI * freq * t) > 0) 1.0 else -1.0
                val sample = square * env
                data[i] = (sample * 16000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }

    fun playSwordSwing() {
        playPcm { rate, _ ->
            val duration = (rate * 0.14).toInt()
            val data = ShortArray(duration)
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val env = sin((t / 0.14) * Math.PI)
                val noise = random.nextDouble() * 2.0 - 1.0
                val sweep = sin(2.0 * Math.PI * (400.0 + (1.0 - t / 0.14) * 800.0) * t)
                val sample = (noise * 0.6 + sweep * 0.4) * env
                data[i] = (sample * 15000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }

    fun playCreeperHiss() {
        playPcm { rate, _ ->
            val duration = (rate * 1.5).toInt()
            val data = ShortArray(duration)
            var last = 0.0
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val env = (t / 1.5).pow(1.5) // escalating volume
                val noise = random.nextDouble() * 2.0 - 1.0
                last = (last * 0.4 + noise * 0.6) // high-pass / noisy hiss
                data[i] = (last * env * 22000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }

    fun playExplosion() {
        playPcm { rate, _ ->
            val duration = (rate * 0.7).toInt()
            val data = ShortArray(duration)
            var rumble = 0.0
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val env = (1.0 - t / 0.7).pow(0.6)
                val noise = random.nextDouble() * 2.0 - 1.0
                rumble = (rumble * 0.85 + noise * 0.15)
                val sub = sin(2.0 * Math.PI * 55.0 * t)
                val sample = ((noise * 0.5 + rumble * 0.3 + sub * 0.2) * env)
                    .coerceIn(-1.0, 1.0)
                data[i] = (sample * 28000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }

    fun playEat() {
        playPcm { rate, _ ->
            val duration = (rate * 0.3).toInt()
            val data = ShortArray(duration)
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val crunchPhase = (t * 8.0) % 1.0
                val env = (1.0 - crunchPhase).pow(2.0)
                val noise = random.nextDouble() * 2.0 - 1.0
                data[i] = (noise * env * 14000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }

    fun playBowShoot() {
        playPcm { rate, _ ->
            val duration = (rate * 0.18).toInt()
            val data = ShortArray(duration)
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val freq = 480.0 - (t / 0.18) * 320.0
                val env = (1.0 - t / 0.18).pow(1.2)
                val sample = sin(2.0 * Math.PI * freq * t) * env
                data[i] = (sample * 18000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }

    fun playCraftSuccess() {
        playPcm { rate, _ ->
            val duration = (rate * 0.35).toInt()
            val data = ShortArray(duration)
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val note1 = sin(2.0 * Math.PI * 523.25 * t) // C5
                val note2 = if (t > 0.12) sin(2.0 * Math.PI * 659.25 * t) else 0.0 // E5
                val note3 = if (t > 0.22) sin(2.0 * Math.PI * 783.99 * t) else 0.0 // G5
                val env = (1.0 - (t / 0.35)).pow(0.8)
                val sample = (note1 * 0.4 + note2 * 0.4 + note3 * 0.5) * env
                data[i] = (sample * 16000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }

    fun playLevelUp() {
        playPcm { rate, _ ->
            val duration = (rate * 0.6).toInt()
            val data = ShortArray(duration)
            val notes = doubleArrayOf(440.0, 554.37, 659.25, 880.0) // A C# E A
            for (i in 0 until duration) {
                val t = i.toDouble() / rate
                val noteIdx = (t / 0.15).toInt().coerceIn(0, 3)
                val freq = notes[noteIdx]
                val noteT = (t % 0.15)
                val env = (1.0 - noteT / 0.15).pow(0.5)
                val sample = sin(2.0 * Math.PI * freq * t) * env
                data[i] = (sample * 18000).toInt().coerceIn(-32767, 32767).toShort()
            }
            data
        }
    }
}

File: js/sound-engine.js
/**
 * Procedural Web Audio API Synthesizer Engine
 * Converted from SoundEngine.kt
 */
class SoundEngine {
  constructor() {
    this.audioCtx = null;
    this.isMuted = false;
    this.sampleRate = 22050;
  }

  initContext() {
    if (!this.audioCtx) {
      const AudioCtxClass = window.AudioContext || window.webkitAudioContext;
      if (AudioCtxClass) {
        this.audioCtx = new AudioCtxClass();
      }
    }
    if (this.audioCtx && this.audioCtx.state === 'suspended') {
      this.audioCtx.resume();
    }
  }

  playPcm(generator) {
    if (this.isMuted) return;
    try {
      this.initContext();
      if (!this.audioCtx) return;

      const floatSamples = generator(this.sampleRate);
      if (!floatSamples || floatSamples.length === 0) return;

      const buffer = this.audioCtx.createBuffer(1, floatSamples.length, this.sampleRate);
      const channelData = buffer.getChannelData(0);
      for (let i = 0; i < floatSamples.length; i++) {
        channelData[i] = floatSamples[i];
      }

      const source = this.audioCtx.createBufferSource();
      source.buffer = buffer;
      source.connect(this.audioCtx.destination);
      source.start();
    } catch (e) {
      // Ignore sound errors gracefully
    }
  }

  playPlace() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 0.08);
      const data = new Float32Array(duration);
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const freq = 240.0 - (t / 0.08) * 120.0;
        const env = Math.pow(1.0 - t / 0.08, 1.5);
        data[i] = Math.sin(2.0 * Math.PI * freq * t) * env * 0.5;
      }
      return data;
    });
  }

  playDig() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 0.06);
      const data = new Float32Array(duration);
      let last = 0.0;
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const env = 1.0 - t / 0.06;
        const noise = Math.random() * 2.0 - 1.0;
        last = last * 0.7 + noise * 0.3;
        data[i] = last * env * 0.45;
      }
      return data;
    });
  }

  playBreak() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 0.15);
      const data = new Float32Array(duration);
      let last = 0.0;
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const env = Math.pow(1.0 - t / 0.15, 0.8);
        const noise = Math.random() * 2.0 - 1.0;
        last = last * 0.6 + noise * 0.4;
        const sub = Math.sin(2.0 * Math.PI * (120.0 - t * 400.0) * t) * 0.5;
        data[i] = (last * 0.8 + sub * 0.2) * env * 0.6;
      }
      return data;
    });
  }

  playStep() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 0.05);
      const data = new Float32Array(duration);
      let last = 0.0;
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const env = 1.0 - t / 0.05;
        const noise = Math.random() * 2.0 - 1.0;
        last = last * 0.8 + noise * 0.2;
        data[i] = last * env * 0.25;
      }
      return data;
    });
  }

  playJump() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 0.12);
      const data = new Float32Array(duration);
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const freq = 160.0 + (t / 0.12) * 200.0;
        const env = Math.sin((t / 0.12) * Math.PI);
        data[i] = Math.sin(2.0 * Math.PI * freq * t) * env * 0.4;
      }
      return data;
    });
  }

  playHurt() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 0.2);
      const data = new Float32Array(duration);
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const freq = 280.0 - (t / 0.2) * 160.0;
        const env = Math.pow(1.0 - t / 0.2, 1.2);
        const square = Math.sin(2.0 * Math.PI * freq * t) > 0 ? 1.0 : -1.0;
        data[i] = square * env * 0.5;
      }
      return data;
    });
  }

  playSwordSwing() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 0.14);
      const data = new Float32Array(duration);
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const env = Math.sin((t / 0.14) * Math.PI);
        const noise = Math.random() * 2.0 - 1.0;
        const sweep = Math.sin(2.0 * Math.PI * (400.0 + (1.0 - t / 0.14) * 800.0) * t);
        data[i] = (noise * 0.6 + sweep * 0.4) * env * 0.45;
      }
      return data;
    });
  }

  playCreeperHiss() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 1.5);
      const data = new Float32Array(duration);
      let last = 0.0;
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const env = Math.pow(t / 1.5, 1.5);
        const noise = Math.random() * 2.0 - 1.0;
        last = last * 0.4 + noise * 0.6;
        data[i] = last * env * 0.65;
      }
      return data;
    });
  }

  playExplosion() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 0.7);
      const data = new Float32Array(duration);
      let rumble = 0.0;
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const env = Math.pow(1.0 - t / 0.7, 0.6);
        const noise = Math.random() * 2.0 - 1.0;
        rumble = rumble * 0.85 + noise * 0.15;
        const sub = Math.sin(2.0 * Math.PI * 55.0 * t);
        data[i] = Math.max(-1.0, Math.min(1.0, (noise * 0.5 + rumble * 0.3 + sub * 0.2) * env)) * 0.8;
      }
      return data;
    });
  }

  playEat() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 0.3);
      const data = new Float32Array(duration);
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const crunchPhase = (t * 8.0) % 1.0;
        const env = Math.pow(1.0 - crunchPhase, 2.0);
        const noise = Math.random() * 2.0 - 1.0;
        data[i] = noise * env * 0.45;
      }
      return data;
    });
  }

  playBowShoot() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 0.18);
      const data = new Float32Array(duration);
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const freq = 480.0 - (t / 0.18) * 320.0;
        const env = Math.pow(1.0 - t / 0.18, 1.2);
        data[i] = Math.sin(2.0 * Math.PI * freq * t) * env * 0.55;
      }
      return data;
    });
  }

  playCraftSuccess() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 0.35);
      const data = new Float32Array(duration);
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const note1 = Math.sin(2.0 * Math.PI * 523.25 * t);
        const note2 = t > 0.12 ? Math.sin(2.0 * Math.PI * 659.25 * t) : 0.0;
        const note3 = t > 0.22 ? Math.sin(2.0 * Math.PI * 783.99 * t) : 0.0;
        const env = Math.pow(1.0 - (t / 0.35), 0.8);
        data[i] = (note1 * 0.4 + note2 * 0.4 + note3 * 0.5) * env * 0.5;
      }
      return data;
    });
  }

  playLevelUp() {
    this.playPcm((rate) => {
      const duration = Math.floor(rate * 0.6);
      const data = new Float32Array(duration);
      const notes = [440.0, 554.37, 659.25, 880.0];
      for (let i = 0; i < duration; i++) {
        const t = i / rate;
        const noteIdx = Math.min(3, Math.floor(t / 0.15));
        const freq = notes[noteIdx];
        const noteT = t % 0.15;
        const env = Math.pow(1.0 - noteT / 0.15, 0.5);
        data[i] = Math.sin(2.0 * Math.PI * freq * t) * env * 0.55;
      }
      return data;
    });
  }
}
